gradoop-flink/src/main/java/org/gradoop/flink/model/api/functions/KeyFunction.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright © 2014 - 2019 Leipzig University (Database Research Group)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradoop.flink.model.api.functions;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+
+import java.io.Serializable;
+
+/**
+ * A function extracting a key from an element. This key may be used as a grouping key.
+ * The key with be an object of a certain type, depending on this function.
+ *
+ * @param <E> The type of the object from which the grouping key is extracted.
+ * @param <K> The type of the extracted key.
+ */
+public interface KeyFunction<E, K> extends Serializable {
+
+  /**
+   * Get the key from the element.
+   *
+   * @param element The element to extract the key from.
+   * @return The key.
+   */
+  K getKey(E element);
+
+  /**
+   * Store a key on an element.<p>
+   * This is used to store the grouping key on the super element after grouping.<p>
+   * The default implementation of this function does not change the element.
+   *
+   * @param element The element where the key should be stored.
+   * @param key     The key to store on the element. The type of the key is expected to be the same as the
+   *                type of keys extracted by this key function.
+   */
+  default void addKeyToElement(E element, Object key) {
```

Comment:
I agree, this is unfortunately not possible since multiple key functions with different key types may be used in one grouping operator instance. Those key functions are stored as a `List<KeyFunction<E, ?>>`. If `K` instead of `Object` would be used there, the method call would be `addKeyToElement(E, ?)` which is not callable with any object.
See

https://github.com/p-f/gradoop/blob/790776ef20897cdf9b27834c3d1b7915e478cff9/gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/keyedgrouping/functions/BuildSuperElementFromTuple.java#L83-L84

for a usage of this function.

Reply:
I thought about the same thing but @p-f is right here - using K instead of Object is not valid